### PR TITLE
Global dismissal and completion animations

### PR DESCRIPTION
**Pull request by creativecreatorormaybenot**
@axel-op I am kinda stuck here.

`markStepComplete` works with the animation now, but `dismiss` gives me a really weird error. Can you check that? I am out of ideas as the implementation for both seems to be the same.

There are a few other minor changes. Maybe you can comment on them if you feel that it is necessary.

### CODE REVIEW

**Comment by axel-op** [comment]
I have entirely refactored the code to implement a stream logic.
It may seems like a lot of useless complexity but I believe that at the end it will be simpler to handle things. 
And I solved the problem (at least I think so, please check yourself).

File: lib/src/enums.dart [hunk 1]
```diff
@@ -0,0 +1,18 @@
+enum DescribedFeatureContentOrientation {
```
Comment (creativecreatorormaybenot) [inline lib/src/enums.dart]: I think that organizing by type is a bad idea. The `enum`s should go where they belong semantically.

File: lib/src/feature_discovery.dart [hunk 2]
```diff
@@ -44,988 +40,130 @@ class FeatureDiscovery extends StatefulWidget {
...
-    _FeatureDiscoveryState.of(context).dismiss();
-  }
-
-  /// This will force clear the current feature discovery and cancel the whole
-  /// process without an animation.
-  /// If you want to dismiss the current step regularly, call [dismiss].
-  static void clear(BuildContext context) {
-    _FeatureDiscoveryState.of(context).clear();
-  }
+  static void dismissCurrentStep(BuildContext context)
```
Comment (creativecreatorormaybenot) [inline lib/src/feature_discovery.dart]: It should be `dismiss` because this does not only dismiss the current step but the whole feature discovery. You should have just kept the naming scheme.

File: lib/src/feature_discovery.dart [hunk 3]
```diff
@@ -44,988 +40,130 @@ class FeatureDiscovery extends StatefulWidget {
...
   /// If the [DescribedFeatureOverlay] that is associated with the current step is
   /// not being displayed, this will fail. In that case, use [clear].
-  static void dismiss(BuildContext context) {
-    _FeatureDiscoveryState.of(context).dismiss();
-  }
-
-  /// This will force clear the current feature discovery and cancel the whole
-  /// process without an animation.
-  /// If you want to dismiss the current step regularly, call [dismiss].
-  static void clear(BuildContext context) {
```
Comment (creativecreatorormaybenot) [inline lib/src/feature_discovery.dart]: Why did you remove this?

File: lib/src/feature_discovery.dart [hunk 4]
```diff
@@ -44,988 +40,130 @@ class FeatureDiscovery extends StatefulWidget {
...
-class _Content extends StatelessWidget {
-  final _OverlayState state;
-  final double transitionProgress;
-  final Offset anchor;
-  final Size screenSize;
-  final double touchTargetRadius;
-
-  // this parameter is not used
-  // final double touchTargetToContentPadding;
+class Bloc {
```
Comment (creativecreatorormaybenot) [inline lib/src/feature_discovery.dart]: BLoC or `Stream`s to be specific is an **alternative** to the `InheritedWidget` model we are using already and mixing them is pointless I think.

File: lib/src/feature_discovery.dart [hunk 5]
```diff
@@ -44,988 +40,130 @@ class FeatureDiscovery extends StatefulWidget {
...
-            fillColor: color,
-            shape: const CircleBorder(),
-            child: child,
-            onPressed: onPressed,
-          ),
-        ),
-      ),
-    );
-  }
+  bool updateShouldNotify(_InheritedFeatureDiscovery old) => old.state.bloc != state.bloc;
```
Comment (creativecreatorormaybenot) [inline lib/src/feature_discovery.dart]: I do not understand 🙁 When should this notify?

File: lib/src/feature_discovery.dart [hunk 6]
```diff
@@ -44,988 +40,130 @@ class FeatureDiscovery extends StatefulWidget {
...
-              height: radius() * 2,
-              decoration: BoxDecoration(
-                shape: BoxShape.circle,
-                color: color.withOpacity(opacity()),
-              ),
-            ),
-          );
-  }
-}
+  final _FeatureDiscoveryState state;
```
Comment (creativecreatorormaybenot) [inline lib/src/feature_discovery.dart]: Why do you pass the `State` down here, should we not just pass the necessary data, so we can also control notifications?

File: lib/src/feature_discovery.dart [hunk 7]
```diff
@@ -1,41 +1,37 @@
...
-  }
-
-  /// It is recommend to use [markStepComplete] whenever you can as it shows an animation for context.
-  ///
-  /// This will force complete the current step and move on to the next step without any animations.
-  static void completeStep(BuildContext context) {
-    _FeatureDiscoveryState.of(context).completeStep();
-  }
+  static void completeCurrentStep(BuildContext context)   
+    => of(context).bloc._inComplete.add(null);
```
Comment (creativecreatorormaybenot) [inline lib/src/feature_discovery.dart]: Even if you wanted to switch to `Stream`s now, it does not seem like a good fit to me since all you are doing is sending `null` events. `InheritedWidget` is good for this because it does what it should: just notify and the event that it notifies can be stored in `bool` values like I did with the sketch.

File: lib/src/feature_discovery.dart [hunk 8]
```diff
@@ -1,41 +1,37 @@
...
-import 'package:flutter/material.dart';
+import 'package:flutter/material.dart' hide OverlayState;
 
 class FeatureDiscovery extends StatefulWidget {
-  const FeatureDiscovery({Key key, this.child}) : super(key: key);
 
-  static String activeStep(BuildContext context) {
-    return _InheritedFeatureDiscovery.of(context).activeStepId;
+  // TODO: This method should not be available to the public
+  static _FeatureDiscoveryState of(BuildContext context) {
```
Comment (creativecreatorormaybenot) [inline lib/src/feature_discovery.dart]: 👍

File: lib/src/described_feature_overlay.dart [hunk 9]
```diff
@@ -0,0 +1,888 @@
+import 'dart:async';
```
Comment (creativecreatorormaybenot) [inline lib/src/described_feature_overlay.dart]: I cannot see what changed here 🙁  It would be beneficial if we changed where the code is located after changing it.

**Review by creativecreatorormaybenot — COMMENTED** [review]
I would revert your changes for now and I think that we can talk about what we should do and where there even are problems before rewriting the whole structure and adding complexity that might not be needed and might introduce problems.

**Comment by creativecreatorormaybenot** [comment]
To me it seems like the `Stream` thing is pulled out of thin air, a bit like you wanted to implement in a way that you know how to do, but maybe there is a very valid reason for why you would want to use `Stream`s.

The current code is not optimal, e.g. the `InheritedWidget` has no function as far as I can tell.

Maybe I should start with what lead you to the decision to switch to BLoC? Or what was wrong with the implementation before specifically?


Here is why I think BLoC is not the optimal solution:

 * `Stream`s are very good for communicating **data**, especially when you have lots of different places, where you consume this data. However, in this case, we do not really want to communicate data, we primarily want to notify and we have exactly one place, where we consume these notifications.

 * The `Stream`s are overly complicated: we need to add additional logic just to handle the `Stream`s and then we need to listen to and send events all over the place. The `InheritedWidget` solution has one place where we notify and another one where we listen. Clean and simple.

 * It does not work with the current implementation. This one is not really an argument against BLoC in this situtation generally, but I will explain why it matters to me in the following.


I really want to get this working and then leave it for now, *now*. I will not be able to work on this at all in the following two weeks and I want to integrate the changes into a production app *today*. Having said that, I will commit a fix to the problem I was having yesterday on this branch because I do not want to use your implementation as it stands.

Your changes will not be lost! I will simply add a new commit and the changes from your commit are still on Git, just not at the head of this branch.  
This version I will use and after that we can see. If you think that your `Stream`s solution really is superior, then go ahead and tinker away, I will be happy to take a look at it.  
Additionally, I would be happy to keep contributing going forward and maybe elevating this package from its current state as it is on Pub. Maybe I could even review some stuff in the next two weeks, but I will not be able to implement something.

I hope that it does not seem like I was being negative towards you - I think that the current implementation your proposed is not polished and I do not have the time to fix it right now. Hopefully my feedback was somewhat useful. I would also love to hear what you think about my implementation because it seems like you just discarded it in favor of this complete (uncomplete, but you know what I mean 🤣 ) change.

**Comment by creativecreatorormaybenot** [comment]
I liked the relocation of code to new files, the example "Cancel" (I will call it "Dismiss") button, and the `of` refactor. I do not know how to extract them from your huge commit, so I will just add them back manually.  
I said that it would be good to not move code and change it in the same commit (because you cannot see the changes), so I will move the code first, commit, and then work on the other changes.

File: lib/src/feature_discovery.dart (hunk 8 above)
Comment (creativecreatorormaybenot) [reply lib/src/feature_discovery.dart]: I just noticed that I misinterpreted this change. I think that it is fine if the method is in `_FeatureDiscoveryState`. Maybe you were unaware of `part` directives, which is the Dart way of handling this.

**Comment by creativecreatorormaybenot** [comment]
There was actually no issue with my code. I accidentally called `dismiss` twice in my app because that was a workaround for the old implementation. I will not merge this for now because it seemed like you did not like the changes I made, but I can use the branch in my project.

**Comment by axel-op** [comment]
It was a starting point. 
As you wrote, my implementation wasn't polished and indeed was far from perfect, but I did this commit (and I probably should have done it elsewhere) in order to propose a new point of view. 

This is the reason why at the moment streams only sent `null` data. But actually I implemented things this way with visibility/privacy in mind, and I thought it could be a way to transmit the data we wanted to without exposing the internal state of our widgets. 

The inherited widget pattern and the BLoC pattern aren't mutually exclusive as you wrote. In fact, they're often used together because an inherited widget is a convenient way to pass the bloc to the other widgets. 

If I implemented streams it wasn't only because I "knew how to do it and wanted to show what I can do" but really because it seems to me that this is the solution to a few issues we have: 

1. As I said before, the bloc pattern helps to hide the internal state of `FeatureDiscovery` by exposing only a few events that will trigger the display/hiding of the other widgets. You probably noticed that I added underscores in front of quite a few fields and variables names. 

2. Consequently it also splits responsibilities: each class should have its own task and shouldn't mind other widgets' business. So while the previous implementation of `DescribedFeatureOverlay` depended a lot on multiple variables inside `FeatureDiscoveryState` (`stepShown`, `activeStepId`, etc.), the bloc pattern here resolves this problem by hiding the variables and only sending the three events `start`, `complete` and `dismiss`. 

3. By splitting responsibility it increases code readability: we don't have anymore to go back and forth between the different files or classes to understand the process. Each class contains its own methods and do not interact with others except with the methods they expose, like an API. 

I improved this first and uncomplete implementation in another branch. Please take a look at it here: https://github.com/axel-op/feature_discovery/tree/with-streams.

File: lib/src/enums.dart (hunk 1 above)
Comment (axel-op) [reply lib/src/enums.dart]: It's true I shouldn't have place them here

File: lib/src/feature_discovery.dart (hunk 7 above)
Comment (axel-op) [reply lib/src/feature_discovery.dart]: I improved it here: https://github.com/axel-op/feature_discovery/tree/with-streams

File: lib/src/feature_discovery.dart (hunk 5 above)
Comment (axel-op) [reply lib/src/feature_discovery.dart]: You're right, I didn't review this part before committing

File: lib/src/feature_discovery.dart (hunk 4 above)
Comment (axel-op) [reply lib/src/feature_discovery.dart]: Not really! they're often used together

File: lib/src/feature_discovery.dart (hunk 3 above)
Comment (axel-op) [reply lib/src/feature_discovery.dart]: This method and the one that 'forces complete' really seem to be duplicate of the regular dismiss and complete. I think that they bring confusion especially in the user's point of view. And I also think that we can make things work without them.

**Comment by creativecreatorormaybenot** [comment]
@axel-op 

> The inherited widget pattern and the BLoC pattern aren't mutually exclusive as you wrote. In fact, they're often used together because an inherited widget is a convenient way to pass the bloc to the other widgets.

I meant the `InheritedWidget` implementation here, i.e. a `StatefulWidget` that is actually responsible for the changes and an `InheritedWidget` that notifies. What you mean is a "passive" `InheritedWidget` because it only provides the BLoC variables or w/e but it is never updated.  
These two do not make a lot of sense in combination I think - and that is the BLoC inside of the `InheritedWidget` or `StatefulWidget` that also notifies, you could indeed have them in a seperate widget.

BLoC, or streams in general for that matter, make more sense when addressing finer parts of the UI, but I still think that it is not necessary here and adds load that is not required. I am still wondering *what is wrong with the current implementation*.

> As I said before, the bloc pattern helps to hide the internal state of FeatureDiscovery by exposing only a few events that will trigger the display/hiding of the other widgets. You probably noticed that I added underscores in front of quite a few fields and variables names.

The `part` directive exists to solve this problem (it is the same as putting the parts in the same file and a file is a library in Dart). You can just add underscores to any variables you like and still reference them internally as long as you declare `part`s. 

> By splitting responsibility it increases code readability: we don't have anymore to go back and forth between the different files or classes to understand the process. Each class contains its own methods and do not interact with others except with the methods they expose, like an API.

With what I wrote before in mind, I think that this just makes things more complicated because there is way more codes and different streams of events and if you were trying to solve the `1.` problem with it, `part`s would be the way to go.

### Conclusion

I cannot tell if I like the new implementatio, I will see when I can take time to look at it and think about it. It would help if you could address what is wrong with the current implementation (again) or respond to what I said because I think that `1.` and `3.` are not problems, they are worse afterwards and `2.` I feel that I will probably agree with you after looking at it, but I am not sure and it seems like that is just personal preference.  
Having said this, I think it'd be best if you just finish what you are thinking because it is probably an improvement and then when I want to contribute in the future, I can just work with it.

**Comment by creativecreatorormaybenot** [comment]
I still think streams are overblown here, but I added some feedback to your streams implementation. 

I am not sure if you are not aware of why we had the widgets previously (what those components do, how they work in the Flutter framework) or you just forgot to adjust it, but you will understand what I mean when you read the comments.
